cli/tools/lint/linter.rs

```diff
@@ -258,3 +317,87 @@ fn apply_lint_fixes(
     deno_ast::apply_text_changes(text_info.text_str(), quick_fixes);
   Some(new_text)
 }
+
+fn run_plugins(
+  plugin_runner: Arc<Mutex<PluginHostProxy>>,
+  parsed_source: ParsedSource,
+  file_path: PathBuf,
+  maybe_token: Option<CancellationToken>,
+) -> Result<ExternalLinterResult, AnyError> {
+  let source_text_info = parsed_source.text_info_lazy().clone();
+  let plugin_info = plugin_runner
+    .lock()
+    .get_plugin_rules()
+    .into_iter()
+    .map(Cow::from)
+    .collect();
+
+  #[allow(clippy::await_holding_lock)]
+  let fut = async move {
+    let mut plugin_runner = plugin_runner.lock();
+    let serialized_ast = plugin_runner.serialize_ast(parsed_source)?;
+
+    plugins::run_rules_for_ast(
+      &mut plugin_runner,
+      &file_path,
+      serialized_ast,
+      source_text_info,
+      maybe_token,
+    )
+    .await
+  }
+  .boxed_local();
+
+  let plugin_diagnostics = tokio_util::create_and_run_current_thread(fut)?;
+
+  Ok(ExternalLinterResult {
+    diagnostics: plugin_diagnostics,
+    rules: plugin_info,
+  })
+}
+
+struct ExternalLinterContainer {
+  cb: Option<ExternalLinterCb>,
+  error: Option<Arc<Mutex<Option<AnyError>>>>,
+}
+
+impl ExternalLinterContainer {
+  pub fn new(
+    maybe_plugin_runner: Option<Arc<Mutex<PluginHostProxy>>>,
+    maybe_token: Option<CancellationToken>,
+  ) -> Self {
+    let mut s = Self {
+      cb: None,
+      error: None,
+    };
+    if let Some(plugin_runner) = maybe_plugin_runner {
+      s.error = Some(Arc::new(Mutex::new(None)));
+      let error_ = s.error.clone();
+      let cb = Arc::new(move |parsed_source: ParsedSource| {
+        let token_ = maybe_token.clone();
+        // TODO: clean this up
+        let file_path = parsed_source.specifier().to_file_path().unwrap();
```

Comment:
Prefer `deno_path_utils::url_to_file_path` and maybe handle these errors?

Reply:
Good point! It will always be a path, so probably error is unreachable.